src/h3lib/lib/geoCoord.c

```diff
@@ -332,3 +324,117 @@ int64_t H3_EXPORT(numHexagons)(int res) {
                                    569707381193162L};
     return nums[res];
 }
+
+/**
+ * Surface area in radians^2 of spherical triangle on unit sphere.
+ *
+ * For the math, see:
+ * https://en.wikipedia.org/wiki/Spherical_trigonometry#Area_and_spherical_excess
+ *
+ * @param   a  length of triangle side A in radians
+ * @param   b  length of triangle side B in radians
+ * @param   c  length of triangle side C in radians
+ *
+ * @return     area in radians^2 of triangle on unit sphere
+ */
+double triangleEdgeLengthsToArea(double a, double b, double c) {
+    double s = (a + b + c) / 2;
+
+    a = (s - a) / 2;
+    b = (s - b) / 2;
+    c = (s - c) / 2;
+    s = s / 2;
+
+    return 4 * atan(sqrt(tan(s) * tan(a) * tan(b) * tan(c)));
+}
+
+/**
+ * Compute area in radians^2 of a spherical triangle, given its vertices.
+ *
+ * @param   a  vertex lat/lng in radians
+ * @param   b  vertex lat/lng in radians
+ * @param   c  vertex lat/lng in radians
+ *
+ * @return     area of triangle on unit sphere, in radians^2
+ */
+double triangleArea(const GeoCoord *a, const GeoCoord *b, const GeoCoord *c) {
+    return triangleEdgeLengthsToArea(H3_EXPORT(pointDistRads)(a, b),
+                                     H3_EXPORT(pointDistRads)(b, c),
+                                     H3_EXPORT(pointDistRads)(c, a));
+}
+
+/**
+ * Area of H3 cell in radians^2.
+ *
+ * The area is calculated by breaking the cell into spherical triangles and
+ * summing up their areas. Note that some H3 cells (hexagons and pentagons)
+ * are irregular, and have more than 6 or 5 sides.
+ *
+ * todo: optimize the computation by re-using the edges shared between triangles
+ *
+ * @param   cell  H3 cell
+ *
+ * @return        cell area in radians^2
+ */
+double H3_EXPORT(cellAreaRads2)(H3Index cell) {
+    GeoCoord c;
+    GeoBoundary gb;
+    H3_EXPORT(h3ToGeo)(cell, &c);
+    H3_EXPORT(h3ToGeoBoundary)(cell, &gb);
+
+    double area = 0.0;
+    for (int i = 0; i < gb.numVerts; i++) {
+        int j = (i + 1) % gb.numVerts;
+        area += triangleArea(&gb.verts[i], &gb.verts[j], &c);
+    }
+
+    return area;
+}
+
+/**
+ * Area of H3 cell in kilometers^2.
+ */
+double H3_EXPORT(cellAreaKm2)(H3Index h) {
+    return H3_EXPORT(cellAreaRads2)(h) * EARTH_RADIUS_KM * EARTH_RADIUS_KM;
+}
+
+/**
+ * Area of H3 cell in meters^2.
+ */
+double H3_EXPORT(cellAreaM2)(H3Index h) {
+    return H3_EXPORT(cellAreaKm2)(h) * 1000 * 1000;
+}
+
+/**
+ * Length of a unidirectional edge in radians.
+ *
+ * @param   edge  H3 unidirectional edge
+ *
+ * @return        length in radians
+ */
+double H3_EXPORT(exactEdgeLengthRads)(H3Index edge) {
```

Comment:
It might be worth extending this to also cover cell perimeter?

Reply:
Sure. Do we think there's a need for that just yet? Does it make sense to add now, or hold off until the need becomes more clear?

Thoughts, @isaacbrodsky @dfellis ?